pkg/k8s/node.go

```diff
@@ -0,0 +1,166 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	PROVIDER_REPORT_MESSAGE_FORMAT         = "%s is unsupported node provider"
+	KERNEL_REPORT_MESSAGE_FORMAT           = "%s is unsupported kernel - minimal: %s"
+	OPERATING_SYSTEM_REPORT_MESSAGE_FORMAT = "%s is unsupported os - only %s supported"
+	CPU_REPORT_MESSAGE_FORMAT              = "insufficient cpu - acutal: %d / minimal: %d"
+	MEMORY_REPORT_MESSAGE_FORMAT           = "insufficient memory - acutal: %dG / minimal: %dG"
+	ARCHITECTURE_REPORT_MESSAGE_FORMAT     = "%s is unsupported architecture - only %s supported"
+)
+
+var (
+	KERNEL_VERSION_REGEX = regexp.MustCompile("^(?P<major>[0-9]).(?P<minor>[0-9]+).(?P<patch>[0-9]+)")
+)
+
+type NodeSummary struct {
+	CPU             int64
+	Memory          int64
+	Name            string
+	Kernel          string
+	Provider        string
+	Architecture    string
+	OperatingSystem string
+}
+
+func (kubeClient *Client) GetNodesSummeries(ctx context.Context) ([]NodeSummary, error) {
+	var err error
+
+	var nodeList *v1.NodeList
+	if nodeList, err = kubeClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{}); err != nil {
+		return nil, err
+	}
+
+	var nodeSummeries []NodeSummary
+	for _, node := range nodeList.Items {
+		nodeSummary := &NodeSummary{
+			Name:            node.ObjectMeta.Name,
+			Provider:        node.Spec.ProviderID,
+			Architecture:    node.Status.NodeInfo.Architecture,
+			Kernel:          node.Status.NodeInfo.KernelVersion,
+			OperatingSystem: node.Status.NodeInfo.OperatingSystem,
+			CPU:             node.Status.Allocatable.Cpu().Value(),
+			Memory:          node.Status.Allocatable.Memory().Value() / (1 << 30),
+		}
+		nodeSummeries = append(nodeSummeries, *nodeSummary)
+	}
+
+	return nodeSummeries, nil
+}
+
+type NodeMinimumRequirements struct {
+	CPUAmount               int64
+	MemoryAmount            int64
+	KernelVersion           semver.Version
+	BlockedProviders        []string
+	AllowedArchitectures    []string
+	AllowedOperatingSystems []string
+	report                  map[string][]string
+}
+
+func NewNodeMinimumRequirements() *NodeMinimumRequirements {
+	return &NodeMinimumRequirements{
+		CPUAmount:               2,
+		MemoryAmount:            4,
+		AllowedOperatingSystems: []string{"linux"},
+		AllowedArchitectures:    []string{"amd64"},
+		BlockedProviders:        []string{"fargate"},
+		KernelVersion:           semver.Version{Major: 4, Minor: 14},
+		report:                  make(map[string][]string),
```

Comment:
If I understand correctly , report contains only unsupported nodes and not a map of each node and a flag of whether it is supported or not, if you intended to hold a report for all the nodes than there is a bug below (supported cluster not included in report) and if not  than rename this field to unsupportedNodesResults

Reply:
Also, a less stateful way to achieve this report is to generate a method IdentifySupportedNodes that receives a NodeSummery list and returns a list of installableNodes  instead of updating an inner field with a state that is initialized after the struct creation, it will be more flexible and more testable